thunderbirdparser/src/org/sleuthkit/autopsy/thunderbirdparser/EmailMessageThreader.java

```diff
@@ -0,0 +1,674 @@
+                if (container.hasMessage()) {
+                    messageID = String.format("<Bogus-id: %d >", bogus_id_count++);
+                    container = null;
+                } else {
+                    container.setMessage(message);
+                }
+            }
+
+            if (container == null) {
+                container = new Container(message);
+                id_table.put(messageID, container);
+            }
+
+            processMessageReferences(message, container, id_table);
+        }
+
+        return id_table;
+    }
+
+    /**
+     * Loops throught message's list of references, creating objects as needed
+     * and setting up the parent child relationships amoung the messages.
+     *
+     * @param message   The current email messags
+     * @param container Container object for message
+     * @param id_table  Hashtable of known message-id\container pairs
+     */
+    void processMessageReferences(EmailMessage message, Container container, Map<String, Container> id_table) {
+        List<String> referenceList = message.getReferences();
+
+        // Make sure the inReplyToID is in the list of references
+        String inReplyToID = message.getInReplyToID();
+        if (inReplyToID != null && !inReplyToID.isEmpty()) {
+            if (referenceList == null) {
+                referenceList = new ArrayList<>();
+            }
+
+            referenceList.add(inReplyToID);
+        }
+
+        // No references, nothing to do
+        if (referenceList == null) {
+            return;
+        }
+
+        Container parent_ref = null;
+        Container ref;
+
+        for (String refID : referenceList) {
+            // Check id_table to see if there is already a container for this
+            // reference id, if not create a new Container and add to table
+            ref = id_table.get(refID);
+
+            if (ref == null) {
+                ref = new Container();
+                id_table.put(refID, ref);
+            }
+
+            // Set the parent\child relationship between parent_ref and ref
+            if (parent_ref != null
+                    && !ref.hasParent()
+                    && parent_ref != ref
+                    && !parent_ref.isChild(ref)) {
+                ref.setParent(parent_ref);
+                parent_ref.addChild(ref);
+            }
+
+            parent_ref = ref;
+        }
+
+        // If the parent_ref and container are already linked, don't change 
+        // anything
+        if (parent_ref != null
+                && (parent_ref == container
+                || container.isChild(parent_ref))) {
+            parent_ref = null;
+        }
+
+        // If container already has a parent, the parent was assumed based on
+        // the list of references from another message.  parent_ref will be
+        // the real parent of container so throw away the old parent and set a
+        // new one.
+        if (container.hasParent()) {
```

Comment:
We should only do this if parent_ref != null, right?

Reply:
No, it took me a bit to wrap my head around this.  I will explain\update the comment.

Reply:
I checked the original code that I based my off of an there is not a check of parent_ref, here are the comments from that code:
```
      // If it has a parent already, that's there because we saw this message
      // in a references field, and presumed a parent based on the other
      // entries in that field.  Now that we have the actual message, we can
      // be more definitive, so throw away the old parent and use this new one.
      // Find this container in the parent's child-list, and unlink it.
      //
      // Note that this could cause this message to now have no parent, if it
      // has no references field, but some message referred to it as the
      // non-first element of its references.  (Which would have been some
      // kind of lie...)
```